README.md
# vscode-plugin
##萌新的练手


mycommand/CHANGELOG.md
# Change Log

All notable changes to the "mycommed" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

mycommand/extension.js
// @ts-nocheck
const vscode = require('vscode');
const datetime = require('silly-datetime');
var cp = require('child_process');



/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {
	//插件控制台
	console.log('mycommend 已启动');
    //hello启动	
	let disposable = vscode.commands.registerCommand('mycommand.SayHello', () => {
		var time=datetime.format(new Date(), 'YYYY-MM-DD HH:mm:ss');
		console.log(time);
		vscode.window.showInformationMessage('欢迎回来'+'('+time+')');
		vscode.window.setStatusBarMessage('插件启动',5000);
	});
    //返回所有命令
    vscode.commands.getCommands().then(allCommands => {
		console.log('所有命令：', allCommands);
	});
	//是否打开网址
	if (vscode.workspace.getConfiguration("mycommandPlugin").get("Promptbox")) {
        vscode.window.showInformationMessage('是否要打开官网？', '是', '否', '不再提示').then(result => {
			if (result === '是') {
				cp.exec('start http://kahxnsat.com/');
			} else if (result === '不再提示') {
				vscode.workspace.getConfiguration().update("mycommandPlugin.Promptbox", false, true);
				vscode.window.showInformationMessage('我呸！');
			}else{
				vscode.window.showInformationMessage("草泥马的");
			}
		});
    }
	
  
	require('./src/cheerio')(context);     //cheerio
	require('./src/getIPAdress')(context); //getipadress 获取本地ip
	require('./src/webview')(context);     //webview 
	require('./src/getpath')(context);     //获取路径
	require('./src/attack')(context);     //attack

	/*插件调用区*/
	context.subscriptions.push(disposable);
}
function deactivate() {}
module.exports = {
	activate,
	deactivate
}

mycommand/src/cheerio.js
const fs=require('fs');
const parse = require('url-parse');
const cheerio=require('cheerio');
const vscode = require('vscode');
module.exports =(context) => {
     //爬虫cheerio
    vscode.commands.registerCommand('mycommand.cheerio',()=>{
		cheerioinput();
    });
};

let cheeriolog = vscode.window.createOutputChannel('CheerioLog');

async function cheerioinput() {
	const result = await vscode.window.showInputBox({
		placeHolder: '请输入要估计的网址（带协议头，如：http:,https:）',
		validateInput: (text) => {
            //判断协议头
            // vscode.window.showInformationMessage(text);
			return (text.substr(0,5)==='http:'||text.substr(0,6)==='https:') ? null : '请输入正确的网址!';
		}
	});
    cheerioweb(result);
	vscode.window.showInformationMessage(`cheerio: ${result}`);
}

function cheerioweb(uri)
{
	cheeriolog.show();
	//要爬取的网址
	// let url='http://www.baidu.com/';
	// let json='http://nodejs.org/dist/index.json';
	const parts = parse(uri);
	const server = parts.protocol == 'https:' ? require('https') : require('http');
	//判断链接
	if(uri===undefined||uri===null){
		vscode.window.showErrorMessage("不能为空");
		return 0;
	}
	//爬虫
	server.get(uri,(res)=>{
		//安全判断
		const {statusCode}=res;            //状态码
		const contentType=res.headers['content-type'];    //文件类型
		cheeriolog.appendLine(statusCode + " " +contentType);
		let err=null;
		if (statusCode!=200) {
			err=new Error("请求状态错误");
		}else if(!/^text\/html/.test(contentType))
		{
			//格式类型是网页文件
			err=new Error("请求类型错误");
		}
		if (err) {
			console.log(err);
			cheeriolog.appendLine('错误');
			res.resume();//重置缓存
			return false;
		}	
		//数据的处理
		//数据分段，只要接受数据就会触发data事件chunk每次接受的数据片段
		let rawData='';
		res.on('data',(chunk)=>{
			// console.log('数据传输');
			cheeriolog.appendLine('------');
			rawData +=chunk.toString('utf8');
			cheeriolog.appendLine(rawData);
		})
		//数据流传输完毕
		res.on('end',(chunk)=>{
			//将请求的数据保存到本地
			// fs.writeFileSync('./baidu.html',rawData);
			fs.writeFileSync('C:/Users/liuzh/Desktop/1.html',rawData);
			cheeriolog.appendLine('数据传输完毕');
			//通过cheerio分析
			let $=cheerio.load(rawData);    //将请求的网页数据进行转化
			$('img').each((index,el)=>{
				cheeriolog.appendLine($(el).attr('src'));
			})
		})
	}).on('error',(err)=>{
		cheeriolog.appendLine('请求错误');
		console.log(err);
	});
}

mycommand/src/webview.js
const vscode = require('vscode');
const path = require('path');
const fs = require('fs');

/**
 * 执行回调函数
 * @param {*} panel 
 * @param {*} message 
 * @param {*} resp 
 */
//调用回调
function invokeCallback(panel, message, resp) {
    console.log('回调消息：', resp);
    // 错误码在400-600之间的，默认弹出错误提示
    if (typeof resp == 'object' && resp.code && resp.code >= 400 && resp.code < 600) {
        vscode.window.showErrorMessage(resp.message || '发生未知错误！');
    }
    panel.webview.postMessage({cmd: 'vscodeCallback', cbid: message.cbid, data: resp});
}

/**
 * 存放所有消息回调函数，根据 message.cmd 来决定调用哪个方法
 */
//消息处理程序
const messageHandler = {
    getConfig(global, message) {
        const result = vscode.workspace.getConfiguration().get(message.key);
        invokeCallback(global.panel, message, result);
    },

    setConfig(global, message) {
        // 写入配置文件，注意，默认写入工作区配置，而不是用户配置，最后一个true表示写入全局用户配置
        vscode.workspace.getConfiguration().update(message.key, message.value, true);
        vscode.window.showInformationMessage('修改配置成功！');
    }
};


/************************************************************************************************************************
 *获取页面函数：                                                                                                           
 *getWebviewContent()                                      直接返回页面内容                                               
 *getWebViewContent(context, templatePath)                 (context,相对路径)
 ************************************************************************************************************************/
//Webview1
function getWebviewContent() {
	return `
    <!DOCTYPE html>
    <html>
    <head>
    	<meta charset="UTF-8">
    	<title>JS实现效果-点击按钮返回到页面顶部</title>
    	<style>
    		#container{margin:0;padding:0;}
    		#box1,#box2,#box3,#box4{width:100%;height:500px;}
    		#box1{background: deepskyblue;}
    		#box2{background: yellowgreen;}
    		#box3{background: darkred;}
    		#box4{background: blueviolet;}
    		#icon{width:50px;height:50px;font-size: 20px;background: #CDCC7D;line-height: 50px;text-align: center;position: fixed;bottom:20px;right:20px;color:#666;font-weight: bolder;}
    	</style>
    </head>
    <body>
    	<div id="container">
    		<div id="box1"></div>
    		<div id="box2"></div>
    		<div id="box3"></div>
    		<div id="box4"></div>   
    		<div id="icon">↑</div>
    	</div>
    </body>
    <script type="text/javascript">
    	window.onload = function(){
    		var btn = document.getElementById("icon");
    		var timer = null;
    		var oScroll = true;
    		//滚动条事件,触发时清空定时器
    		window.onscroll = function(){
    			if(!oScroll){
    				clearInterval(timer);
    			}
    			oScroll = false;
    		}
    		btn.onclick = function(){
    			//加入定时器让他又快到慢滚动到顶部
    			timer = setInterval(function(){
    				//获取当前scrollTop的高度位置（兼容ie和chrom浏览器）
    				var oTop = document.documentElement.scrollTop || document.body.scrollTop;
    				//设置速度由快到慢
    				var ispeed = Math.floor(-oTop / 7);
    				document.documentElement.scrollTop = document.body.scrollTop = oTop + ispeed;
    				oScroll = true;
    				if(oTop == 0){
    					clearInterval(timer);
    				}
    			},30);
    		}
    	}
    </script>
    </html>`;
}
//webview2
function getWebViewContent(context, templatePath) {
	const resourcePath = path.join(context.extensionPath, templatePath);
	console.log(resourcePath);
	let html = fs.readFileSync(resourcePath, 'utf-8');
    // // vscode不支持直接加载本地资源，需要替换成其专有路径格式，这里只是简单的将样式和JS的路径替换
	// const dirPath = path.dirname(resourcePath);	
	// html = html.replace(/(<link.+?href="|<script.+?src="|<img.+?src=")(.+?)"/g, (m, $1, $2) => {
	// 	return $1 + vscode.Uri.file(path.resolve(dirPath, $2)).with({ scheme: 'vscode-resource' }).toString() + '"';
	// });
	return html;
}

module.exports=(context) => {

	context.subscriptions.push(vscode.commands.registerCommand('mycommand.webview.hello', ()=>{
		const panel=vscode.window.createWebviewPanel(
			'webview',  //webview内部识别
			'未定义欢迎页',  //显示页面名
			vscode.ViewColumn.One,    //现在在编辑器那个部位
			{// Webview选项
				enableScripts: true, // 启用JS，默认禁用
				retainContextWhenHidden: true, // webview被隐藏时保持状态，避免被重置
			}		
		);
        let global = { panel};
        // 设置HTML内容
		// panel.webview.html=getWebviewContent();	
		panel.webview.html = getWebViewContent(context,"./src/source/webview.html");

        // 监听
        panel.webview.onDidReceiveMessage(message => {
            //更新数据
            if (messageHandler[message.cmd]) {
                console.log(message.cmd);
                messageHandler[message.cmd](global, message);
            } else {
                vscode.window.showErrorMessage(`未找到名为 ${message.cmd} 回调方法!`);
            }
        }, undefined, context.subscriptions);
    }));
    // 如果设置里面开启了欢迎页显示，启动欢迎页
    if (vscode.workspace.getConfiguration("mycommandPlugin").get("ShowTip")) {
        vscode.commands.executeCommand('mycommand.webview.hello');  //执行命令
    }
}






mycommand/src/getIPAdress.js
const vscode = require('vscode');
const os = require('os');
var internetAvailable = require("internet-available");
//获取本地ip 函数
function getIPAddress(){
    // var interfaces = os.networkInterfaces();
    // for(var devName in interfaces){
    //     var iface = interfaces[devName];
    //     for(var i=0;i<iface.length;i++){
    //         var alias = iface[i];
    //         if(alias.family === 'IPv4' && alias.address !== '127.0.0.1' && !alias.internal){
    //             return alias.address;
    //         }
    //     }
    // }

	var ifaces = os.networkInterfaces();
		Object.keys(ifaces).forEach((ifname)=> {
		    var alias = 0;
		    ifaces[ifname].forEach((iface)=>{
			    if ('IPv4' !== iface.family || iface.internal !== false) {
			    //跳过内部地址（即127.0.0.1）和非ipv4地址
			    return '127.0.0.1';
			    }
    
			    if (alias >= 1) {
			    //这个单一介面有多个ipv4地址
			    vscode.window.showInformationMessage(ifname + ':' + alias, iface.address);
			    console.log(ifname + ':' + alias, iface.address);
			    } else {
			    //该接口只有一个ipv4地址
			    vscode.window.showInformationMessage(ifname, iface.address);
			    console.log(ifname, iface.address);
			    }
			++alias;
		});
	});
}

//获取本地ip
module.exports = (context) => {
    context.subscriptions.push(vscode.commands.registerCommand('mycommand.getIPAdress', ()=> {
		// vscode.window.showInformationMessage('local IP: '+ getIPAddress()); // 本地ip
	 var hostName=os.hostname(); 
		vscode.window.showInformationMessage('local host:'+ hostName);
		getIPAddress();

		//判断网络状况
		internetAvailable(/*{
			timeout: 4000,      //检测超时时间
			retries: 10,        //检测次数
		   }*/).then(function(){
			console.log("Internet available");
		   }).catch(function(){
			console.log("No internet");
		});
		
		//internet-available默认检测的DNS域名是google.com，如果想自定义域名，代码如下：
		// internetAvailable({
		// 	domainName: "xxxxx.com",
		// 	port: 53,
		// 	host: '8.8.8.8' // 默认，国内请改成114.114.114.114
		//    }).then(() => {
		// 	console.log("Internet available");
		//    }).catch(() => {
		// 	console.log("No internet");
		// });
    }));
};


mycommand/src/attack.js
const vscode = require('vscode');
const parse = require('url-parse');
const chalk = require('chalk');

module.exports=(context) => {
vscode.commands.registerCommand('mycommand.attack', ()=> {   
    AttackUrl();
  })
}

//输出面板
// let Attacklog = vscode.window.createOutputChannel('Attacklog');//在输出的任务选项下创建一个输出面板
let Attacklog = vscode.window.createOutputChannel('AttackLog');//在输出的任务选项下创建一个输出面板

//攻击网址输入
async function AttackUrl() {
	const result = await vscode.window.showInputBox({
		placeHolder: '请输入要估计的网址（带协议头，如：http:,https:）',
		validateInput: (text) => {
      //判断协议头
      // vscode.window.showInformationMessage(text);
			return (text.substr(0,5)==='http:'||text.substr(0,6)==='https:') ? null : '请输入正确的网址!';
		}
	});
  Attackhttp(result);
	vscode.window.showInformationMessage(`attack: ${result}`);
}

//attack
function Attackhttp(uri){
    const parts = parse(uri);
    const server = parts.protocol == 'https:' ? require('https') : require('http');
    const runs = [1, 10, 10, 100, 100, 100, 200, 200, 200, 200, 500, 1000, 10000]    
    /**
     * 对HTTP / HTTPS端点进行压力测试，并增加曲线
     *
     * 使用方式: attack https://url.com/foo
     */
    if(uri===undefined||uri===null){
      vscode.window.showErrorMessage("不能为空");
      return 0;
    }
     //显示输出面板
    Attacklog.show();                        
    Attacklog.appendLine("attack:" + uri);
    // console.log(parts);
    function attack (count, cb) {
      let good = 0
      let bad = 0
      let requests = []
      for (let n = 0; n < count; n++) {
        requests.push(
          server.get({
            hostname: parts.hostname,
            port: parts.port,
            agent: false
          }, function (res) {
            good++
            if (good + bad == count) {
              cb(good, bad)
            }else{
                console.log(res);
            }
          })
          .on('error', function(err) {
            bad++
            //如果超过20％的请求出错，则保释
            if (bad/(good+bad) > 0.2) {
              // cancel in flight requests
              requests.forEach(function (request) {
                if (request.abort) {
                  request.abort()
                }
              })
              cb(good, bad)
            }
            if (good + bad == count) {
              cb(good, bad)
            }else{
                // console.log(err);
                Attacklog.appendLine("error");
                throw err;
            }
          })
        )
      }
    }  
    function go (runs) {
      let start = Date.now()
      attack(runs[0], function (good, bad) {
        let total = good + bad;
        let rate = good/total;
        let time = Date.now() - start;
        let niceRate = (100*rate).toFixed(2) + '%';
        if (rate == 1) {
          niceRate = chalk.green(niceRate)
        } else if (rate > 0.9) {
          niceRate = chalk.yellow(niceRate)
        } else {
          niceRate = chalk.red(niceRate)
        }
        // console.log(chalk.blue(runs[0], 'requests:', good + '/' + total, '(' + niceRate + ') - done in', time + 'ms (' + (time/total).toFixed(2), 'ms/request, ' + (total/(time/1000)).toFixed(2) + 'qps)'));
        Attacklog.appendLine(chalk.blue(runs[0], 'requests:', good + '/' + total, '(' + niceRate + ') - done in', time + 'ms (' + (time/total).toFixed(2), 'ms/request, ' + (total/(time/1000)).toFixed(2) + 'qps)'));
        if (rate > 0.8) {
          setTimeout(function(){
            go(runs.slice(1))
          }, 5000);
    
        } else {
          Attacklog.appendLine('done!')
        }

      })
    }
    Attacklog.appendLine('swarming:'+ uri + '...');
    go(runs)
}